File: src/lib/api/installer.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type {
  InstallerEnvironment,
  InstallerRunResult,
  ManualInstallCommandGroup,
  InstallExecutionStep,
} from "@/types/installer";

export const installerApi = {
  detectEnvironment(): Promise<InstallerEnvironment> {
    return invoke("detect_installer_environment");
  },

  installMissing(): Promise<InstallerRunResult> {
    return invoke("install_missing_dependencies");
  },

  getManualCommands(): Promise<ManualInstallCommandGroup[]> {
    return invoke("get_manual_install_commands");
  },

  async subscribeProgress(
    handler: (event: InstallExecutionStep) => void,
  ): Promise<UnlistenFn> {
    return listen<InstallExecutionStep>("installer-progress", ({ payload }) => {
      handler(payload);
    });
  },
};

File: src-tauri/src/commands/installer.rs
#![allow(non_snake_case)]

use tauri::AppHandle;

#[tauri::command]
pub async fn detect_installer_environment(
) -> Result<crate::services::installer::types::InstallerEnvironment, String> {
    Ok(crate::services::installer::detect::detect_installer_environment())
}

#[tauri::command]
pub async fn install_missing_dependencies(
    app: AppHandle,
) -> Result<crate::services::installer::install::InstallerRunResult, String> {
    crate::services::installer::install::install_missing_dependencies(&app).await
}

#[tauri::command]
pub async fn get_manual_install_commands(
) -> Result<Vec<crate::services::installer::install::ManualInstallCommandGroup>, String> {
    Ok(crate::services::installer::install::get_manual_install_commands(
        std::env::consts::OS,
    ))
}
